### PR TITLE
perf: improve the performance of interpodaffinity.Filter

### PR DESCRIPTION
**Pull request by binacs**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

May be able to achieve small performance improvements.

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:

Prepend Label's judgment and postpend AffinityTerm.Matches, which has a higher computational complexity.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
NONE
```

### CODE REVIEW

File: pkg/scheduler/framework/plugins/interpodaffinity/filtering.go [hunk 1]
```diff
@@ -102,21 +102,19 @@ func (m topologyToMatchedTermCount) clone() topologyToMatchedTermCount {
 }
 
 func (m topologyToMatchedTermCount) update(node *v1.Node, tk string, value int64) {
```
Comment (sanposhiho) [inline pkg/scheduler/framework/plugins/interpodaffinity/filtering.go:104]: Add comment like 

```
update updates ....
This function assumes the node has the label keyed with `tk`.
```

File: pkg/scheduler/framework/plugins/interpodaffinity/filtering.go [hunk 2]
```diff
@@ -149,6 +153,11 @@ func podMatchesAllAffinityTerms(terms []framework.AffinityTerm, pod *v1.Pod) boo
 	return true
 }
 
+func nodeInterestedInTopology(node *v1.Node, topologyKey string) bool {
```
Comment (sanposhiho) [inline pkg/scheduler/framework/plugins/interpodaffinity/filtering.go:156]: something like `nodeHasTopologyKeyLabel` looks more straightforward.

File: pkg/scheduler/framework/plugins/interpodaffinity/filtering.go (hunk 1 above)
Comment (binacs) [reply pkg/scheduler/framework/plugins/interpodaffinity/filtering.go:106]: fixed, thanks for review!

File: pkg/scheduler/framework/plugins/interpodaffinity/filtering.go (hunk 2 above)
Comment (binacs) [reply pkg/scheduler/framework/plugins/interpodaffinity/filtering.go]: fixed

File: pkg/scheduler/framework/plugins/interpodaffinity/filtering.go [hunk 3]
```diff
@@ -101,22 +101,22 @@ func (m topologyToMatchedTermCount) clone() topologyToMatchedTermCount {
...
+// This function assumes the node has the label keyed with `tk`.
 func (m topologyToMatchedTermCount) update(node *v1.Node, tk string, value int64) {
-	if tv, ok := node.Labels[tk]; ok {
-		pair := topologyPair{key: tk, value: tv}
-		m[pair] += value
-		// value could be negative, hence we delete the entry if it is down to zero.
-		if m[pair] == 0 {
-			delete(m, pair)
-		}
+	pair := topologyPair{key: tk, value: node.Labels[tk]}
```
Comment (sanposhiho) [inline pkg/scheduler/framework/plugins/interpodaffinity/filtering.go:107]: Sorry, on the second thought, can we change `update` fn to receive the label value like:

```
func (m topologyToMatchedTermCount) update(topologykey, topologyvalue string, value int64) {
```

Then, we reduce the num of lookup `node.Labels[tk]` here. Also, we can eliminate the future mistake of passing `node` which doesn't have the label keyed with `tk`

Comment (binacs) [reply pkg/scheduler/framework/plugins/interpodaffinity/filtering.go]: Nice catch!

Changed the `update` function and lookup the `node.Labels` as little as possible, PTAL

File: pkg/scheduler/framework/plugins/interpodaffinity/filtering.go [hunk 4]
```diff
@@ -101,24 +101,24 @@ func (m topologyToMatchedTermCount) clone() topologyToMatchedTermCount {
...
-func (m topologyToMatchedTermCount) update(node *v1.Node, tk string, value int64) {
-	if tv, ok := node.Labels[tk]; ok {
-		pair := topologyPair{key: tk, value: tv}
-		m[pair] += value
-		// value could be negative, hence we delete the entry if it is down to zero.
-		if m[pair] == 0 {
-			delete(m, pair)
-		}
+// update updates the count of pods that meet the matching criteria in a particular topology.
+// This function assumes the node has the label keyed with `tk`.
```
Comment (sanposhiho) [inline pkg/scheduler/framework/plugins/interpodaffinity/filtering.go:105]: 2nd line of the comment is no longer needed.

**Review by sanposhiho — COMMENTED** [review]
Thanks, one nit only; LGTM otherwise

**Comment by binacs** [comment]
kindly ping @sanposhiho s

File: pkg/scheduler/framework/plugins/interpodaffinity/filtering.go [hunk 5]
```diff
@@ -128,18 +127,25 @@ func (m topologyToMatchedTermCount) updateWithAffinityTerms(
...
 // returns true IFF the given pod matches all the given terms.
-func podMatchesAllAffinityTerms(terms []framework.AffinityTerm, pod *v1.Pod) bool {
+func podMatchesAllAffinityTerms(node *v1.Node, terms []framework.AffinityTerm, pod *v1.Pod) bool {
 	if len(terms) == 0 {
 		return false
 	}
 	for _, t := range terms {
+		if _, ok := node.Labels[t.TopologyKey]; !ok {
+			continue
+		}
```
Comment (sanposhiho) [inline pkg/scheduler/framework/plugins/interpodaffinity/filtering.go:148]: Why do we need to have this change?
We've confirmed that node has all TopologyKey label in L349 - L359.

Comment (binacs) [reply pkg/scheduler/framework/plugins/interpodaffinity/filtering.go:148]: `podMatchesAllAffinityTerms` will be called not only in `satisfyPodAffinity`, but also in `updateWithAffinityTerms` (the latter does not ensure that the label must exist)

**Comment by binacs** [comment]
reping @sanposhiho

**Pull request closed without merging**